### ISSUE TITLE
authz/github: repo-centric groups sync does not account for repo being visibility `internal`

### ISSUE THREAD

**Issue by bobheadxi** [sourcegraph/sourcegraph#25904]
**This pertains to [`groupsCacheTTL`-enabled sync](https://docs.sourcegraph.com/admin/repo/permissions#teams-and-organizations-permissions-caching) only.**

We don't currently honor GitHubs internal respositories, instead treating them as private, where they should actually be open to anyone that's a member of the GitHub org.

The [current implementation](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c11a86a7194a00b73be4895ec0dec1c9ade2a3b8/-/blob/enterprise/internal/authz/github/github.go#L503:20) only grants access to:
- users with direct affiliations
- the entire org, if org-level visibility is read
- affiliated teams, if org-level visibility is not read

This needs to check for another case: if repo visibility is `internal`, we should grant access to the entire org as well

Note that this is **not** related to some of the other `internal`-visibility-related issues like the one outlined here: https://github.com/sourcegraph/sourcegraph/issues/17153#issuecomment-852846420

**Comment by indradhanush** [comment]
I've been involved with managing incidents and working on their follow ups this week and haven't been able to dedicate time to this issue. The above linked discussion in the meantime has opened up some new questions which might help us in finding a solution. Next steps are to follow up on the discussion and perform discovery on feasible solutions.

**Comment by indradhanush** [comment]
An update on what's been brewing off late:

Earlier today we found out what appears to be a way to build a relatively easier solution without having to design an elaborate pipeline for syncing internal repos exclusively. 

The plan is documented [here](https://github.com/sourcegraph/sourcegraph/discussions/26837#discussioncomment-1611792) in the GitHub discussion where I documented the challenges and suggestions.

#27272 is a work in progress for the first stage of this plan. 

Also a heads up, I'm on leave on Wednesday and Thursday this week and I'll be at the Berlin Individual Meetup the whole of next week. This will slow us down somewhat. I'm hoping to be able to get a working solution out sometime next week. I'll post an update with a more accurate timeline once I make a little more headway into the work.

**Comment by indradhanush** [comment]
#### Update 

#27272 is close to completion (pending some additional tests). Post which I will move on to step 2 of the proposed plan [here](https://github.com/sourcegraph/sourcegraph/discussions/26837#discussioncomment-1611792):

1. Modify the GitHub client to support the visibility field - Done when #27272 is merged.
2. Write visibility field to repos.metadata during repo sync
3. Query metadata field in code path that determines visibility

**Comment by indradhanush** [comment]
**Update 2021-11-22**

During the implementation of step 1 of the plan documented above, we ran into a **blocker**. It turns out that the `visibility` field is only supported in GitHub Enterprise 3.3.0 for the GraphQL API although the support is available form 2.2.0 for the REST API. For the implementation to work correctly, we will need support from both the GraphQL and the REST API as we end up using both in different code paths. This is required to guarantee that we always have the correct `visibility` metadata available for each repository. And GitHub Enterprise 3.3.0 is a release candidate as of this writing and most likely that our customers won't be running this version when we roll the fix out.

As a result we will have to either:
1. Require that customers run GitHub Enterprise 3.3.0 to avail the fix
2. Change our plan to make an additional REST API call to sync the repo `visibility` metadata. This would ensure that the `visibility` metadata is always up to date for every repo and our system can rely on it with strong guarantees. However, this would come at the cost of an extra API call and to proceed with this plan, I will need to discuss this with the team to come up with a feasible plan of action.

Also, in the meantime we've been asked to help with a [P1 customer issue](https://github.com/sourcegraph/customer/issues/559) and we will need to shift focus this week to this issue to keep up with the SLAs of P1 issues. As a result, we are pausing work on this issue for the time being.

**Comment by kkom** [comment]
@indradhanush – just note that some Sourcegraph users (e.g. us) are on GitHub Enterprise Cloud – so it's not necessary a case of requiring them to upgrade on a tight schedule.

PS: We're not using `groupsCacheTTL` anyway, so waiting for the other side of https://github.com/sourcegraph/sourcegraph/issues/21591 to be addressed. But I assume a similar consideration will come up.

**Comment by indradhanush** [comment]
**Update 2021-11-25**

Now that the P1 customer issue was resolved, we've moved focus back to this. We made some new changes to the work in #27272. 

The next step after it is merged would be to move to step 2. Simultaneously we will co-ordinate with @sourcegraph/delivery to upgrade our GHE instance to 3.3.0. 

I will update this issue when we have more visibility into the upgrade of our internal GHE instance as it would be a blocker for us to test all the work that will follow next.

**Comment by indradhanush** [comment]
**Update 2021-12-15**

Our GHE instance has been upgraded to 3.3.0 and changes in #29013 validates the correctness of the work we did in #27272. 

Once this is merged, we will proceed with the [next steps](https://github.com/sourcegraph/sourcegraph/issues/25904#issuecomment-973990734) sequentially:

> 2. Write visibility field to repos.metadata during repo sync
> 3. Query metadata field in code path that determines visibility

**Comment by bobheadxi** [comment]
> Write visibility field to repos.metadata during repo sync
Query metadata field in code path that determines visibility

Shouldn't granting access to all org users in github authz have the same affect, rather than leaking the abstraction out of github authz into general authz and having to be aware of visibility elsewhere? It also seems like it would be a safer implementation

**Comment by indradhanush** [comment]
>  Shouldn't granting access to all org users in github authz have the same affect and be a safer implementation, rather than leaking the abstraction out of github authz into general authz and having to be aware of visibility elsewhere?

@bobheadxi Yes that's correct, but that breaks the UX of `internal` repos for users on GitHub. They want an easy way to just mark the repo as `internal` which makes it accessible for everyone in the org by default. And at the moment we're treating `internal` repos the same as `private` repos, which is what this would then fix. 

Did I misunderstand your question by any chance? :thinking: 

Happy to chat on Slack and get more clarification if you'd like before I go ahead with the next step.

**Comment by indradhanush** [comment]
**Update 2021-12-21**

I had a chat with @bobheadxi on Slack last week and here is what we have agreed upon as the best way forward:

The logic for determining `authz` based access **should not** be leaked into `repo.metadata` as planned [initially](https://github.com/sourcegraph/sourcegraph/discussions/26837#discussioncomment-1611792). This is because, if we go ahead with this and update the code path to determine if the user can view the repo or not [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/authz/github/client.go?L12:6&subtree=true), then we will also have to update every touch point where the repo / user permissions table is accessed and query for `repo.metadata` to determine the repo's visibility. This makes it a bad hack and leaves the state of repo permissions in a brittle state. 

Instead, we will go ahead and incur an extra API call to get the repo's visibility in [getRepoAffiliatedGroups](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7348499e5b703896fc0714e6f015d4a87b926902/-/blob/enterprise/internal/authz/github/github.go?L542&subtree=true) to determine if the repo is `internal` or not.

Additionally, this allows us to be always have the latest visibility state of the repo in question instead of relying on the visibility from the last time the repo sync. This will avoid a time lag between when a repo's visibility is changed to Sourcegraph making the right `authz` decision. 

This change of plan appears last minute, but the detail was only evident on further thought and discussion and we're confident that the cost of the extra API call here is the right step to take with the bigger picture in mind. This feature *will be* guarded behind a feature flag to avoid potential regressions.

**Comment by indradhanush** [comment]
#### Update 2022-01-27

#29325 introduces a fix to handle internal repo visibility. This is guarded behind a feature flag `experimentalFeatures.enableGithubInternalRepoVisibility` and turned off by default. 

Next step is a follow up work to verify if user-centric permissions sync can overwrite the repo-centric permissions sync in the case of internal repos. If this turns out to be true, we will need to follow up with some additional changes in the user-centric permissions sync.

**Comment by indradhanush** [comment]
We had to revert #29325 because of a failing test in main. Will look into this next.
